### PR TITLE
Fix #8685 - added systemd sd_notify implementation

### LINKED ISSUE #8685: Notify systemd about successful startup
**Use case**
I have a service which starts after `clickhouse-server.service` startup, but `clickhouse-server.service` considered started right after fork called by systemd (Type=simple).
This causes connection errors to database at server startup.
I tried to use option `--daemon` and `Type=forking`, but clickhouse forks before start listening http ([Http Interface](https://clickhouse.yandex/docs/en/interfaces/http/)).

**Describe the solution you'd like**
Using [systemd/sd-notify](https://www.freedesktop.org/software/systemd/man/sd_notify.html) will allow us to using `Type=notify`.
I suggest to sent notification if `--notify` arg specified .

**Describe alternatives you've considered**
Refactoring `--daemon` arg, to fork after database is ready to use.

### PR DESCRIPTION
**Pull request by socketpair**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- Systemd integration now correctly notifies systemd that service is really started and is ready to server requests.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: programs/server/Server.cpp [hunk 1]
```diff
@@ -646,6 +650,44 @@ static void sanityChecks(Server & server)
     }
 }
 
+#if defined(OS_LINUX)
+static void _sd_notify(const char*command)
```
Comment (socketpair) [inline programs/server/Server.cpp:654]: I'm not a C++ developer, so made in C style. It works, but looks like something outdated.

**Comment by socketpair** [comment]
```
INFO:root:Getting URL with 5 tries and sleep 0 in between: https://api.github.com/repos/ClickHouse/ClickHouse/pulls/43400
INFO:root:Getting URL with 5 tries and sleep 0 in between: https://api.github.com/users/socketpair/orgs
INFO:root:Getting URL with 5 tries and sleep 0 in between: https://github.com/ClickHouse/ClickHouse/pull/43400.diff
Fetched info about 2 changed files
Got labels set()
INFO:root:User 'socketpair' is not trusted
PRs by untrusted users need the 'can be tested' label - please contact a member of the core team
INFO:botocore.credentials:Found credentials from IAM Role: ec2_admin
Change labels: add ['pr-improvement'], remove []
Notice: Cannot run
```

@alexey-milovidov @den-crane 

I'm trusted :)

**Comment by socketpair** [comment]
@den-crane ubsan detected a bug, that I did not make.

```
Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6) Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6) Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6) Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6) Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6) Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6) Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6) Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6) Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6) Logical error: 'Bad cast from type DB::ColumnNullable to DB::ColumnLowCardinality'. Received signal 6 Received signal Aborted (6)
```

It's definitely not the problem with this PR.

**Comment by socketpair** [comment]
Regarding Clang Tidy:

image


The function IS thread-safe.

**Comment by socketpair** [comment]
regarding other checks:

```
INFO:root:Exception '429 Client Error: too many requests for url: https://api.github.com/repos/ClickHouse/ClickHouse/actions/runs/3507981378/jobs?per_page=100' while getting, retry 1
INFO:root:Exception '429 Client Error: too many requests for url: https://api.github.com/repos/ClickHouse/ClickHouse/actions/runs/3507981378/jobs?per_page=100' while getting, retry 2
INFO:root:Exception '429 Client Error: too many requests for url: https://api.github.com/repos/ClickHouse/ClickHouse/actions/runs/3507981378/jobs?per_page=100' while getting, retry 3
INFO:root:Exception '429 Client Error: too many requests for url: https://api.github.com/repos/ClickHouse/ClickHouse/actions/runs/3507981378/jobs?per_page=100' while getting, retry 4
```

**Comment by socketpair** [comment]
@den-crane well, @davenger has made code better. What's next ?

**Comment by davenger** [comment]
AST Fuzzer (msan) failure: https://github.com/ClickHouse/ClickHouse/issues/43202

**Review by davenger — APPROVED** [review]

**Pull request merged**